qiskit/optimization/algorithms/optimization_algorithm.py

```diff
@@ -104,24 +104,54 @@ class OptimizationResult:
         results: The original results object returned from the optimization algorithm. This can
             contain more information than only the optimal value and function value.
         status: The termination status of the algorithm.
+        variables: The list of variables under optimization.
     """
 
     Status = OptimizationResultStatus
 
     def __init__(self, x: Optional[Any] = None, fval: Optional[Any] = None,
                  results: Optional[Any] = None,
-                 status: OptimizationResultStatus = OptimizationResultStatus.SUCCESS) -> None:
-        self._val = x
+                 status: OptimizationResultStatus = OptimizationResultStatus.SUCCESS,
+                 variables: Optional[List[Variable]] = None) -> None:
+        self._x = x  # pylint: disable=invalid-name
+        self._variables = variables
         self._fval = fval
         self._results = results
         self._status = status
 
     def __repr__(self):
-        return '([{}] / {} / {})'.format(','.join([str(x_) for x_ in self.x]), self.fval,
-                                         self.status)
+        return 'optimal variables: [{}]\noptimal function value: {}\nstatus: {}' \
+            .format(','.join([str(x_) for x_ in self.x]), self.fval, self.status.name)
 
-    def __str__(self):
-        return 'x=[{}], fval={}'.format(','.join([str(x_) for x_ in self.x]), self.fval)
+    def __getitem__(self, item: Union[int, str]):
+        if isinstance(item, int):
+            return self.x[item]
+        if isinstance(item, str):
+            return self.variables[item]
+        raise QiskitOptimizationError("Integer or string parameter required, instead "
+                                      + type(item).__name__ + " provided.")
+
+    @property
+    def variables(self) -> Optional[Dict[str, int]]:
+        """Returns the pairs of variable names and values under optimization.
+
+        Returns:
+            The pairs of variable names and values under optimization.
+        """
+        if self.x is not None and self.variables is not None and len(self.x) == len(self.x_names):
+            return dict(zip(self.x_names, self.x))
+        return None
```

Comment:
These lines are called every time when I access result by a var name, e.g. `result["var_name"]`. The dictionary is recreated every time.

Reply:
It is due to `self.x` can be optional by the current design. I would like make it mandatory parameter of the constructor.

Reply:
I think OptimizationResult should be designed as an immutable object style (readonly, only constructor can set values, parameters of constructors are mandatory). What do you think?

Reply:
@t-imamichi Totally agree! Immutable object with a bunch of read-only properties.

Reply:
Thanks. I will work on it later.